Doc/cnf conformance yml usage #537 (#549)
cncf/cnf-conformance: CNF_CONFORMANCE_YML_USAGE.md

### Table of Contents
- [Overview](#Overview-of-all-cnf-conformance.yml)
- [Keys and Values](#Keys-and-Values)
    - [helm_directory](#helm_directory)
    - [git_clone_url](#git_clone_url)
    - [install_script](#install_script)
    - [release_name](#release_name)
    - [deployment_name](#deployment_name)
    - [deployment_label](#deployment_label)
    - [service_name](#service_name)
    - [application_deployment_name](#application_deployment_name)
    - [docker_repository](#docker_repository)
    - [helm_repository](#helm_repository)
    - [helm_chart](#helm_chart)
    - [helm_chart_container_name](#helm_chart_container_name)
    - [white_list_helm_chart_container_names](#white_list_helm_chart_container_names)
    - [container_names](#container_names)
- [Creating Your Own cnf-conformance.yml](#creating-your-own-cnf-conformanceyml)
- [Detailed Steps](#Detailed-Steps)
- [Setup and Configuration](#Setup-and-Configuration)
- [Quick Setup and Config Reference Steps](#Quick-Setup-and-Config-Reference-Steps)

### Overview of all cnf-conformance.yml

> The following is a basic example cnf-conformance.yml file that can be found in the cnf-conformance respository: [cnf-conformance.example.yml](https://github.com/cncf/cnf-conformance/blob/develop/cnf-conformance.example.yml)

Author: We should probably update this example with the required minimum keys/values.

Reviewer: @agentpoyo good idea. that can be a new PR and/or issue

  - name: sidecar-container2
    rolling_update_test_tag: "1.32.0"
```
### Keys and Values

> #### helm_directory

Reviewer 2: I would just have the bit about mutual exclusivity **bolded**

This is the path to the helm chart directory (relative to the location of the cnf-conformance.yml). This or [helm_chart](#helm_chart) must be set, but only one **(mutually exclusive)**.

Used for doing static tests on the helm chart code e.g. searching for Hardcoded IPs.

An example of a helm chart source directory can be found [here](https://github.com/helm/charts/tree/master/stable/coredns).

> The PATH is also relative to the location of the cnf-conformance.yml. So if the cnf-conformance.yml in the directory ```charts/stable/cnf-conformance.yml``` and helm_directory is set to ```helm_directory: coredns``` the test suite would expect to find the chart under [```charts/stable/coredns```](https://github.com/helm/charts/tree/master/stable/coredns)

Reviewer: ```suggestion
The PATH is also relative to the location of the cnf-conformance.yml. So if the cnf-conformance.yml in the directory ```cnfs/coredns/cnf-conformance.yml``` and helm_directory is set to ```helm_directory: coredns``` the test suite would expect to find the chart under [```cnfs/coredns/coredns```](https://github.com/helm/charts/tree/master/stable/coredns)
```

We refer to the directory cnfs/cnf_name repeatedly in the docs.  It is confusing to see the cnf-conformance.yml in what looks like a helm charts directory (charts/stable)

Author: Yes. I left these as is from the old doc. Wasn't sure if we wanted to entertain changing, updating, coming up with a new standard generic way of explaining the structure.

This is used by the CNF-Conformance suite to interact with the Helm release / installation of the CNF being tested and find meta-data about the CNF.

For example, the [rolling_update](https://github.com/cncf/cnf-conformance/blob/96cee8cefc9a71e62e971f8f4abad56e5db59866/src/tasks/configuration_lifecycle.cr#L156) test uses the helm release_name to fetch the docker image name and tag of the CNF so it can preform a rolling update. [See: rolling_update test](https://github.com/cncf/cnf-conformance/blob/96cee8cefc9a71e62e971f8f4abad56e5db59866/src/tasks/configuration_lifecycle.cr#L179)

#### deployment_name

Example setting: 

`deployment_name: coredns-coredns`

#### deployment_label

Example setting:

`deployment_label: k8s-app`

#### service_name

Example setting:

`service_name: coredns-coredns`

#### application_deployment_name

Example setting:

`application_deployment_names: [coredns-coredns]`

#### docker_repository

Example setting:

`docker_repository: coredns/coredns`

#### helm_repository
This is the URL of your helm repository for your CNF.

Example setting:

```yaml=
helm_repository:
  name: stable 
  repo_url: https://cncf.gitlab.io/stable
```

#### helm_chart

> The published helm chart name. Like [helm_directory](#helm_directory), this or [helm_directory](#helm_directory) must be set, but not both (mutually exclusive).

Reviewer 2: I would just have the bit about mutual exclusivity **bolded**

  - name: sidecar-container2
    rolling_update_test_tag: "1.32.0"
```
### Keys and Values

#### helm_directory

> This is the path to the helm chart directory (relative to the location of the cnf-conformance.yml). This is a must and is mutually exclusive with [helm_chart](#helm_chart)

Reviewer: ```suggestion
This is the path to the helm chart directory (relative to the location of the cnf-conformance.yml). This or   [helm_chart](#helm_chart) must be set, but only one (mutually exclusive).
```

Example setting:

`install_script: cnfs/coredns/Makefile`

#### release_name
This is the helm release name of the CNF.

If the CNF isn't pre-deployed to the cluster then the test suite will perform the installation and use this name for the helm release / version.

This MAY be set.  If release_name is not set, a release name will be generated. 

Example setting:

`release_name: privileged-coredns`

This is used by the CNF-Conformance suite to interact with the Helm release / installation of the CNF being tested and find meta-data about the CNF.

For example, the [rolling_update](https://github.com/cncf/cnf-conformance/blob/96cee8cefc9a71e62e971f8f4abad56e5db59866/src/tasks/configuration_lifecycle.cr#L156) test uses the helm release_name to fetch the docker image name and tag of the CNF so it can preform a rolling update. [See: rolling_update test](https://github.com/cncf/cnf-conformance/blob/96cee8cefc9a71e62e971f8f4abad56e5db59866/src/tasks/configuration_lifecycle.cr#L179)

> #### deployment_name

Author: If someone wants to help add more descriptions for this and the few others that have been added since this was originally written, that would be great.

Reviewer: @williscool @denverwilliams @wavell

Reviewer: @agentpoyo we can do a new PR in the future when @williscool @denverwilliams @wavell add descriptions

Example setting: 

`deployment_name: coredns-coredns`

#### deployment_label

Example setting:

`deployment_label: k8s-app`

#### service_name

Example setting:

`service_name: coredns-coredns`

#### application_deployment_name

Example setting:

`application_deployment_names: [coredns-coredns]`

#### docker_repository

Example setting:

`docker_repository: coredns/coredns`

#### helm_repository
This is the URL of your helm repository for your CNF.

Example setting:

```yaml=
helm_repository:
  name: stable 
  repo_url: https://cncf.gitlab.io/stable
```

#### helm_chart

> The published helm chart name. Like [helm_directory](#helm_directory), this is a must and is exclusive with helm_directory.

Reviewer: ```suggestion
The published helm chart name. Like [helm_directory](#helm_directory), this or [helm_directory](#helm_directory) must be set, but not both (mutually exclusive).
```

Exmple setting:

`helm_chart: stable/coredns`

An example of a publishe helm chart repo/image can be found [here](https://github.com/helm/charts/tree/master/stable/coredns#tldr).

#### helm_chart_container_name

This value is the name of the 'container' defined in the Kubernetes pod spec of the CNF being tested.

> This value is used to look up the CNF and determine if it's running in privileged mode (only used within the specs).  The containers in the test are now dynamically determined from the helm chart or manifest files (See: ['privileged' test](https://github.com/cncf/cnf-conformance/blob/c8a2d8f06c5e5976acd1a641350978929a2eee12/src/tasks/security.cr#L32)).

Reviewer: @agentpoyo maybe this should not be here... or there can be a **Development** section at the end with these items hidden.

Author: :thinking: 

I think all options for a config file should be mentioned since it's listing out all of it's 'usage' within. If we start dividing it up, it really starts to get convoluted or perhaps missed...

Example setting:

`helm_chart_container_name: privileged-coredns-coredns`

#### white_list_helm_chart_container_names

> This key and value is the name of the 'container' defined in the Kubernetes pod spec of pods that are allowed to be running in privileged mode. (Optional)

Reviewer: ```suggestion
The values of this key are the names of the 'containers' defined in the Kubernetes pod spec of pods that are allowed to be running in privileged mode. (Optional)
```

This value is used to allow 'particular' pods to run in privileged mode on the K8s cluster where is CNF being tested is installed.
The reason this is needed is because the Test Suite will check, 'all' pods in the cluster, to see if they're running in privileged mode.

This is done because it's a common cloud-native practice to delegate 'privileged' networking tasks to only a single app e.g Multus, NSM vs making the CNF privileged itself. As a consequence the whitelist can only be used to exempt 'privileged' infrastructure services running as pods e.g NSM, Multus and cannot be used to exempt the CNF being tested.

Example setting:

`white_list_helm_chart_container_names: [coredns]`

#### container_names

This value is the name of the 'containers' defined in the Kubernetes pod spec of pods and must be set. 

Example setting:

```yaml=
container_names: #[LIST_OF_CONTAINERS_NAMES_AND_VERSION_UPGRADE_TAGS]
   - name: <container_name1>
     rolling_update_test_tag: <image-tag-version1>
   - name: <container_name2>
     rolling_update_test_tag: <image-tag-version2>
```

> 

Reviewer: NOTE/TODO: This should be dynamic or it will be difficult for CNFs with a large number of containers.

(No update to docs at this time)

Author: Yes, I left this as is from original doc.

This value is used to test the upgradeability of each container image.  The image tag version should be a minor version that will be used in conjunction with the kubnetes rollout feature.

> ### Creating Your Own cnf-conformance.yml

Author: This was leftovers from the INSTALL, not sure if we need to keep or not since it kind of dives into setup below as well.

> ### Detailed Steps

Reviewer: ```suggestion
```

Delete it

Author: I'll remove and cleanup the TOC and such from command line.

> This will help guide you in setting up your own cnf-conformance.yml to test your own CNF.

Reviewer: ```suggestion
A configuration file called `cnf-conformance.yml` needs to be created for each CNF you want to test (eg. `cnfs/my_ipsec_cnf/cnf-conformance.yml`).
```

> You can start by copying the example cnf-conformance.yml or copy and paste the below to get started and then filling our the appropriate values:

Reviewer: ```suggestion
You can start by copying an example cnf-conformance.yml or copy and paste the below to get started and then filling our the appropriate values:
```
